driver-bitfury16.c

```diff
@@ -0,0 +1,4589 @@
+
+	info->chipboard[board_id].bcm250[bcm250_id].chips[chip_id].errors++;
+
+	if ((renonce_chip(chip_address) == 0) ||
+	    (opt_bf16_renonce == RENONCE_DISABLED))
+		info->chipboard[board_id].bcm250[bcm250_id].chips[chip_id].status = UNINITIALIZED;
+
+	/* mark chip as FAILING if error rate too high*/
+	if (info->chipboard[board_id].bcm250[bcm250_id].chips[chip_id].error_rate >= CHIP_ERROR_LIMIT) {
+#ifdef FILELOG
+		filelog(info, "BF16: chip [%d:%d:%2d] error rate too high: [%d], "
+				"marked as failing, recovery_count: [%d]",
+				board_id, bcm250_id, chip_id,
+				info->chipboard[board_id].bcm250[bcm250_id].chips[chip_id].error_rate,
+				info->chipboard[board_id].bcm250[bcm250_id].chips[chip_id].recovery_count);
+#endif
+
+		info->chipboard[board_id].bcm250[bcm250_id].chips[chip_id].status = FAILING;
+		info->chipboard[board_id].bcm250[bcm250_id].chips[chip_id].error_rate = 0;
+	}
+}
+
+static uint8_t* init_channel_path(uint8_t board_id, uint8_t btc250_num, uint8_t* channel_depth, uint8_t channel_length)
+{
+	uint8_t i, j;
+	uint8_t* channel_path = cgcalloc(channel_length, sizeof(uint8_t));
+
+	for (i = 0, j = 0; i < CHANNEL_DEPTH; i++) {
+		if (bcm250_map[board_id][btc250_num].channel_path[i] != BF250_NONE)
+			(*channel_depth)++;
+
+		int8_t shift = 8*(j + 1) - 3*(i + 1);
+		if (shift < 0) {
+			channel_path[j]     |= bcm250_map[board_id][btc250_num].channel_path[i] >> abs(shift);
+			channel_path[j + 1] |= bcm250_map[board_id][btc250_num].channel_path[i] << (8 - abs(shift));
+			j++;
+		} else
+			channel_path[j] |= bcm250_map[board_id][btc250_num].channel_path[i] << shift;
+	}
+
+	return channel_path;
+}
+
+static int8_t parse_chip_address(struct cgpu_info *bitfury, char* address, bf_chip_address_t* chip_address)
+{
+	struct bitfury_info *info = (struct bitfury_info *)(bitfury->device_data);
+
+	uint8_t board_id  = 0;
+	uint8_t bcm250_id = 0;
+	uint8_t chip_id   = 0;
+
+	char buff[16];
+
+	if (address == NULL)
+		return -1;
+
+	/* board_id */
+	char* start = strchr(address, '[');
+	char* end   = strchr(address, ':');
+	if ((start == NULL) || (end == NULL))
+		return -1;
+	uint8_t len = end - start;
+
+	memset(buff, 0, sizeof(buff));
+	cg_memcpy(buff, start + 1, len);
+	board_id = atoi(buff);
+
+	/* bcm250_id */
+	start = end;
+	end = strchr(start + 1, ':');
+	if (end == NULL)
+		return -1;
+	len = end - start;
+
+	memset(buff, 0, sizeof(buff));
+	cg_memcpy(buff, start + 1, len);
+	bcm250_id = atoi(buff);
+
+	/* chip_id */
+	start = end;
+	end = strchr(start + 1, ']');
+	if (end == NULL)
+		return -1;
+	len = end - start;
+
+	memset(buff, 0, sizeof(buff));
+	cg_memcpy(buff, start + 1, len);
+	chip_id = atoi(buff);
+
+	if ((board_id < 0) || (board_id >= CHIPBOARD_NUM)) {
+		applog(LOG_ERR, "%s: invalid board_id %d: [0 - %d] specified",
+				bitfury->drv->name,
+				board_id, CHIPBOARD_NUM);
+		return -1;
+	}
+
+	if ((bcm250_id < 0) || (bcm250_id >= BCM250_NUM)) {
```

Comment:
driver-bitfury16.c:677:17: warning: comparison is always false due to limited range of data type [-Wtype-limits]
  if ((bcm250_id < 0) || (bcm250_id >= BCM250_NUM)) {